src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt

```diff
@@ -5,81 +5,89 @@ import com.refinedmods.refinedstorage.api.network.node.INetworkNodeFactory
+
+            val nodesTag: ListTag = tag.getList(NBT_NODES, Constants.NBT.COMPOUND_TAG)
+
             nodes.clear()
-            for (i in nodesTag.indices) {
+            nodesTag.indices.forEach { i ->
                 val nodeTag: CompoundTag = nodesTag.getCompound(i)
                 val id = Identifier(nodeTag.getString(NBT_NODE_ID))
                 val data: CompoundTag = nodeTag.getCompound(NBT_NODE_DATA)
                 val pos: BlockPos = BlockPos.fromLong(nodeTag.getLong(NBT_NODE_POS))
-                val factory: INetworkNodeFactory? = API.networkNodeRegistry.get(id)
-                if (factory != null) {
-                    var node: INetworkNode? = null
-                    try {
-                        node = factory.create(data, world, pos)
-                    } catch (t: Throwable) {
-                        logger.error("Could not read network node", t)
-                    }
-                    if (node != null) {
-                        nodes[pos] = node
+
+                when(val factory: INetworkNodeFactory? = API.networkNodeRegistry[id]){
+                    null -> log.warn("Factory for $id not found in network node registry")
+                    else -> {
+                        try {
+                            nodes[pos] = factory(data, world, pos)
+                        } catch (t: Throwable) {
+                            log.error("Could not read network node", t)
+                        }
                     }
-                } else {
-                    logger.warn("Factory for $id not found in network node registry")
                 }
             }
         }
     }
 
     override fun toTag(tag: CompoundTag): CompoundTag {
         val list = ListTag()
-        for (node in all()) {
+        nodes.values.forEach { node ->
             try {
                 val nodeTag = CompoundTag()
                 nodeTag.putString(NBT_NODE_ID, node.id.toString())
                 nodeTag.putLong(NBT_NODE_POS, node.pos.asLong())
                 nodeTag.put(NBT_NODE_DATA, node.write(CompoundTag()))
                 list.add(nodeTag)
             } catch (t: Throwable) {
-                logger.error("Error while saving network node", t)
+                log.error("Error while saving network node", t)
             }
         }
         tag.put(NBT_NODES, list)
         return tag
     }
 
+    override fun getCachedNode(pos: BlockPos): INetworkNode? = nodes[pos]
+
     override fun getNode(pos: BlockPos): INetworkNode? {
         // TODO possible self-registering cache alternative
-//        // Use cache if it exists
-//        if(nodes.contains(pos)) {
-//            return nodes[pos]
-//        }
-//
 //        // Lookup and cache
 //        return BlockComponents.get(RSComponents.NETWORK_NODE_PROXY, world, pos)?.let {
 //            nodes[pos] = it.node
 //            it.node
 //        }
 
+        // Use cache if it exists
+        if(nodes.contains(pos)) {
```

Comment:
Maybe. It might be best to move the `markedForRemoval` flag into the node and check that. I was avoiding re-looking up the entity via the world (which would defeat the point of the cache.)

Reply:
This has been updated, but could do with another review.